feat: Add Papercut module

TechLiam opened the pull request:
## What does this PR do?
Add Testcontainers.Papercut

## Why is it important?
Papercut is a testing SMTP server that also has a nice, easy REST API to get the messages in the inbox.
When doing integration testing it is helpful to be able to send emails and have the ability to check their contents by having this container developers will be able to quickly spin up an SMTP server and check the messages delivered to it without having to send emails to a proper SMTP server and email address.

## How to test this PR
Unit/Integration tests have been added to the solution that checks the container starts and can have emails sent to it. It also checks the helper methods for getting the emails delivered back out of Papercut from the REST API.

TechLiam committed:
Added Papercut module

TechLiam committed:
Improved unit test so its not time dependant but will still time out

TechLiam committed:
Added the extra API calls

TechLiam committed:
Removed using that not needed

HofmeisterAn commented on src/Testcontainers.Papercut/PapercutContainer.cs:
```diff
@@ -0,0 +1,76 @@
+        var client = new HttpClient();
+        client.BaseAddress = new Uri(WebUrl);
+        return client.GetAsync($"/api/messages/{id}/contents/{contentId}");
+    }
```
Is there no Papercut client or, at the very least, a dependency that holds the domain models? Is there a publicly available API document we can refer to? The Testcontainers modules do not implement client functionalities; instead, we focus on spinning up dependencies for various use cases. Can we remove these client functionalities? We can move and keep the necessary domain models to the test and simply send the required HTTP requests from there.

TechLiam replied:
Cool, yeah, that makes sense responsibility wise all the container package is doing is creating the container, and interacting with it is outside the scope of this package. I have removed the HTTP Client calls from the package code (moved some/most of it to the tests)

HofmeisterAn commented on src/Testcontainers.Papercut/PapercutBuilder.cs:
```diff
@@ -0,0 +1,71 @@
+    protected override PapercutBuilder Init()
+    {
+        return base.Init()
+            .WithImage("jijiechen/papercut:latest")
```
Please use a const and pin the image version, as is done with all the other modules.

TechLiam replied:
Moved it to a const not sure how to pin the version as it's not tagged in the docker hub had a look at how the other packages are doing this and could see that some of the packages e.g. Testcontainers.CosmosDb are using :latest

I'm reasonably new to docker stuff though so if there is a way to do this please let me know I'm happy to add it in

HofmeisterAn replied:
> not sure how to pin the version as it's not tagged

👍 Okay, then `latest` is fine. I didn't check the image tags. Thanks for checking.

HofmeisterAn commented on src/Testcontainers.Papercut/PapercutBuilder.cs:
```diff
@@ -0,0 +1,71 @@
+        return base.Init()
+            .WithImage("jijiechen/papercut:latest")
+            .WithPortBinding(25,true)
+            .WithPortBinding(37408,true)
```
Please use a const and a meaningful name that describes the port.

TechLiam replied:
Done

HofmeisterAn commented on src/Testcontainers.Papercut/PapercutBuilder.cs:
```diff
@@ -0,0 +1,71 @@
+using System.Net;
```
Move all usings to the global use file.

TechLiam replied:
Done

HofmeisterAn commented on tests/Testcontainers.Papercut.Tests/PapercutContainerTest.cs:
```diff
@@ -0,0 +1,47 @@
+    public async Task SendingAnEmail()
+    {
+        //Given
+        var client = new SmtpClient(_papercutContainer.Hostname,_papercutContainer.GetMappedPublicPort(25));
```
WDYT about a delegate to access the SmtpPort?

TechLiam replied:
I agree this is better so have added it in

HofmeisterAn reviewed (changes requested):
Thanks for your PR. Definitely a cool module. Can we please remove the Papercut client functionalities? This isn't the responsibility of Testcontainers.

TechLiam committed:
Removed the client calls as this is out of scope for the package
Moved client models to test project as this uses them but it not used in package project now
Added constants for Port numbers
Moved using statements to Usings.cs file
Added delegate for getting public SMTP port number

TechLiam committed:
Merge branch 'develop' into develop

HofmeisterAn committed:
Merge branch 'develop' of github.com:TechLiam/testcontainers-dotnet into TechLiam-develop

TechLiam committed:
Merge branch 'develop' into develop

TechLiam commented:
@HofmeisterAn is there anything else or is this PR good to merge now?

HofmeisterAn committed:
chore: Align Papercut implementation

HofmeisterAn committed:
Merge branch 'develop' of github.com:TechLiam/testcontainers-dotnet into TechLiam-develop

HofmeisterAn commented:
> @HofmeisterAn is there anything else or is this PR good to merge now?

I've made some improvements. WDYT? If you're good with it, we are ready to merge.

TechLiam commented on tests/Testcontainers.Papercut.Tests/PapercutContainerTest.cs:
```diff
@@ -16,33 +16,57 @@ public Task DisposeAsync()
+        Assert.Equal(subject, messages[0].Subject);
+    }
+
+    private readonly struct Message
```
I like this struct hadn't thought about doing it with that good call

HofmeisterAn reviewed (dismissed):
I guess you're good with the recent changes. Thanks for the pull request. Great first Testcontainers contribution 👍. I will merge it later the day.

HofmeisterAn committed:
chore: Remove using
